Add current_time Presto function

minhancao opened the pull request:
PR to implement current_time function for https://github.com/facebookincubator/velox/issues/4466

minhancao commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1041,4 +1042,33 @@ struct ParseDateTimeFunction {
+
+    using namespace std::chrono;
+
+    auto now = system_clock::now();
```
Will try that out, thanks!

minhancao replied:
@pedroerp I have changed the implementation to use `Timestamp::now()` and also the DateTimeFormatter that is in Velox

aditi-pandit replied:
@minhancao : It might be cleaner to add a function in Timestamp to print out the iso8601 string representation with timezone and invoke that with Timestamp::now().

I also didn't understand the purpose of the formatter here.

minhancao commented:
Hi @pedroerp and @aditi-pandit, I've been trying to run my test that I written in this PR and I've been getting the following error output:
```
[ RUN      ] DateTimeFunctionsTest.currentTimeTest
E20230420 13:13:14.318490 8818080 Exceptions.h:68] Line: /Users/minhancao/velox/velox/parse/TypeResolver.cpp:108, Function:resolveScalarFunctionType, Expression:  Scalar function doesn't exist: get_current_time., Source: USER, ErrorCode: INVALID_ARGUMENT
unknown file: Failure
C++ exception with description "Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Scalar function doesn't exist: get_current_time.
Retriable: False
Function: resolveScalarFunctionType
File: /Users/minhancao/velox/velox/parse/TypeResolver.cpp
Line: 108
```

It looks like it's looking for "get_current_time" which I'm not sure why since I supplied the call with "current_time" in my test... if there is any insight on this please let me know, thanks!

aditi-pandit reviewed (commented):
Thanks @minhancao for adding this function.

minhancao commented:
Hi @majetideepak, thank you for the suggestion! I have the following code so far for what you've suggested:
```
core::TypedExprPtr currentTimeCall() {
    return std::make_shared<core::CallTypedExpr>(
        VARCHAR(), args, "current_time");
  }
  ```

Since current_time doesn't take in any args, I'm not sure what to put here for the args parameter, if you have insight on this please let me know! And also I'm not sure where I should put this piece of code, should it be in the test file? 
  
current_time would be ran like the following query:
```
presto:tpch> select current_time;
      _col0       
------------------
 16:34:03.409 UTC 
(1 row)

Query 20230402_125419_00003_8w2rp, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
[Latency: client-side: 202ms, server-side: 160ms] [0 rows, 0B] [0 rows/s, 0B/s]
```

majetideepak commented:
@minhancao try an empty vector for the args. Yes, this code will be in the test itself.

minhancao commented:
@majetideepak Thank you, that worked for building my expression. This is what I have so far in DateTimeFunctionsTest.cpp:
```
core::TypedExprPtr currentTimeCall() {
    std::vector<facebook::velox::core::TypedExprPtr> emptyVectors;
    return std::make_shared<core::CallTypedExpr>(
        VARCHAR(), emptyVectors, "current_time");
  }
  ```
  
  ```
  TEST_F(DateTimeFunctionsTest, currentTimeTest) {
  // test the built expression
  auto expression =
      currentTimeCall();

  ASSERT_EQ("current_time", compile(expression)->toString());

  const auto current_time = [&]() {   
    return evaluateOnce<std::string>("current_time");
  };

  std::optional<std::string> currentTimeStr = current_time();

  EXPECT_EQ("19:51:34.241 UTC", currentTimeStr);
}
```

When this piece of code is ran:
```
const auto current_time = [&]() {   
    return evaluateOnce<std::string>("current_time");
  };
  ```
  , it is still being routed to duckdb "get_current_time"... should I be testing in a different way or am I missing something here?
  
  ```
  E20230502 09:46:15.805547 10894078 Exceptions.h:68] Line: /Users/minhancao/velox/velox/parse/TypeResolver.cpp:108, Function:resolveScalarFunctionType, Expression:  Scalar function doesn't exist: get_current_time., Source: USER, ErrorCode: INVALID_ARGUMENT
unknown file: Failure
C++ exception with description "Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Scalar function doesn't exist: get_current_time.
Retriable: False
Function: resolveScalarFunctionType
File: /Users/minhancao/velox/velox/parse/TypeResolver.cpp
Line: 108
```

minhancao commented:
Hi @majetideepak, I tried your suggestion and got the following error:
```
/Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2895:77: error: no matching member function for call to 'makeNullableFlatVector'
    auto result = evaluate(expression, makeRowVector(std::vector<VectorPtr>{makeNullableFlatVector()}));
                                                                            ^~~~~~~~~~~~~~~~~~~~~~
/Users/minhancao/velox/./velox/vector/tests/utils/VectorTestBase.h:169:30: note: candidate function template not viable: requires at least argument 'data', but no arguments were provided
  FlatVectorPtr<EvalType<T>> makeNullableFlatVector(
                             ^
1 error generated.
```

This is what I have for the test:
```
TEST_F(DateTimeFunctionsTest, currentTimeTest) {
    auto expression =  currentTimeCall();
    auto result = evaluate(expression, makeRowVector(std::vector<VectorPtr>{makeNullableFlatVector()}));
    auto currentTimeStr = castEvaluateResult<std::string>(result, expression->toString());

    EXPECT_EQ("19:51:34.241 UTC", currentTimeStr);
}
```

aditi-pandit commented:
> auto result = evaluate(expression, makeRowVector(std::vector<VectorPtr>{makeNullableFlatVector()}));

Hi @minhancao :
Function makeNullableFlatVector is a templated function that requires a type variable for its instantiation. In your case it seems like you just need a bunch of rows to instantiate this function multiple times (to ensure it returns the same value each time) 
auto result = evaluate(expression, makeRowVector({makeNullableFlatVector<int32_t>(
      {10, 10, std::nullopt, 15})}));
Please can you check if that works.

minhancao commented:
Hi @aditi-pandit, I have tried what you suggested and now I'm getting these errors:
```
/Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2897:27: error: 'castEvaluateResult' is a private member of 'facebook::velox::functions::test::FunctionBaseTest'
    auto currentTimeStr = castEvaluateResult<std::string>(result, expression->toString());
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/minhancao/velox/./velox/functions/prestosql/tests/utils/FunctionBaseTest.h:241:22: note: declared private here
  std::shared_ptr<T> castEvaluateResult(
                     ^
In file included from /Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:19:
In file included from /Users/minhancao/velox/./velox/common/base/tests/GTestUtils.h:19:
/Users/minhancao/velox/_build/debug/_deps/gtest-src/googletest/include/gtest/gtest.h:1374:11: error: invalid operands to binary expression ('const char[17]' and 'const std::shared_ptr<std::string>')
  if (lhs == rhs) {
      ~~~ ^  ~~~
/Users/minhancao/velox/_build/debug/_deps/gtest-src/googletest/include/gtest/gtest.h:1393:12: note: in instantiation of function template specialization 'testing::internal::CmpHelperEQ<char[17], std::shared_ptr<std::string>>' requested here
    return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
           ^
/Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2899:5: note: in instantiation of function template specialization 'testing::internal::EqHelper::Compare<char[17], std::shared_ptr<std::string>, nullptr>' requested here
    EXPECT_EQ("19:51:34.241 UTC", currentTimeStr);
    ```

minhancao commented:
I have made some changes to my implementation of the function in DateTimeFunctions.h, please review and let me know if I need to edit anything, thanks!

aditi-pandit commented:
@minhancao : Please can you check the errors in the build, format-check.

aditi-pandit commented on velox/docs/functions/presto/datetime.rst:
```diff
@@ -24,6 +24,10 @@ Date and Time Functions
 
     Returns ``timestamp`` as a UNIX timestamp.
 
+.. function:: current_time() -> string
```
The functions are added in the documentation in alphabetical order. Please add this function in the correct order in this file.

minhancao replied:
@aditi-pandit Moved it to where it should be in alphabetical order

aditi-pandit commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -226,6 +228,20 @@ class DateTimeFunctionsTest : public functions::test::FunctionBaseTest {
+
+  core::TypedExprPtr currentTimeCall() {
+    std::vector<facebook::velox::core::TypedExprPtr> emptyVectors;
+    return std::make_shared<core::CallTypedExpr>(
```
Is it possible to just put {} inline instead of creating a variable for emptyVectors ?

minhancao replied:
@aditi-pandit I tried your suggestion `return std::make_shared<core::CallTypedExpr>(VARCHAR(), {}, "current_time");` and it errored out with the following when I tried to build:
```
[1/2] Building CXX object velox/functions/prestosql/tests/CMakeFiles/velox_functions_test.dir/DateTimeFunctionsTest.cpp.o
FAILED: velox/functions/prestosql/tests/CMakeFiles/velox_functions_test.dir/DateTimeFunctionsTest.cpp.o 
ccache /Library/Developer/CommandLineTools/usr/bin/c++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DFMT_LOCALE -DFOLLY_HAVE_INT128_T=1 -DGFLAGS_IS_A_DLL=0 -I/Users/minhancao/velox/. -I/Users/minhancao/velox/velox/external/xxhash -I/Users/minhancao/velox/_build/debug/_deps/xsimd-src/include -I/Users/minhancao/velox/_build/debug -isystem /usr/local/include -isystem /Users/minhancao/velox/velox -isystem /Users/minhancao/velox/velox/external -isystem /Users/minhancao/velox/velox/external/duckdb -isystem /Users/minhancao/velox/velox/external/duckdb/tpch/dbgen/include -isystem /usr/local/opt/openssl@1.1/include -isystem /Users/minhancao/velox/_build/debug/_deps/gtest-src/googletest/include -isystem /Users/minhancao/velox/_build/debug/_deps/gtest-src/googletest -isystem /Users/minhancao/velox/_build/debug/_deps/gtest-src/googlemock/include -isystem /Users/minhancao/velox/_build/debug/_deps/gtest-src/googlemock -isystem /Users/minhancao/velox/_build/debug/_deps/gtest-src/include -isystem /Users/minhancao/velox/_build/debug/_deps/gtest-src -mavx2 -mfma -mavx -mf16c -mlzcnt -std=c++17 -mbmi2 -fvisibility=hidden -fvisibility-inlines-hidden -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -Wall -Wextra -Wno-unused        -Wno-unused-parameter        -Wno-sign-compare        -Wno-ignored-qualifiers        -Wno-range-loop-analysis          -Wno-mismatched-tags          -Wno-nullability-completeness -Werror -g -std=gnu++17 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk -fPIE -MD -MT velox/functions/prestosql/tests/CMakeFiles/velox_functions_test.dir/DateTimeFunctionsTest.cpp.o -MF velox/functions/prestosql/tests/CMakeFiles/velox_functions_test.dir/DateTimeFunctionsTest.cpp.o.d -o velox/functions/prestosql/tests/CMakeFiles/velox_functions_test.dir/DateTimeFunctionsTest.cpp.o -c /Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp
/Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:241:12: error: no matching function for call to 'make_shared'
    return std::make_shared<core::CallTypedExpr>(
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__memory/shared_ptr.h:1113:17: note: candidate template ignored: substitution failure [with _Tp = facebook::velox::core::CallTypedExpr]: deduced incomplete pack <std::shared_ptr<const ScalarType<TypeKind::VARCHAR> >, (no value), const char (&)[13]> for template parameter '_Args'
shared_ptr<_Tp> make_shared(_Args&& ...__args)
                ^
1 error generated.
```

aditi-pandit replied:
That's fine. Thanks !

pedroerp replied:
make_shared does not support non-explicit initializer lists, which is super annoying :(

minhancao commented:
@aditi-pandit I have addressed your PR comments and implemented as you suggested in Timestamp.h, please take a look when you can, thanks!

Also I was wondering if you have any insight on this error below:

> Hi @aditi-pandit, I have tried what you suggested and now I'm getting these errors:
> 
> ```
> /Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2897:27: error: 'castEvaluateResult' is a private member of 'facebook::velox::functions::test::FunctionBaseTest'
>     auto currentTimeStr = castEvaluateResult<std::string>(result, expression->toString());
>                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /Users/minhancao/velox/./velox/functions/prestosql/tests/utils/FunctionBaseTest.h:241:22: note: declared private here
>   std::shared_ptr<T> castEvaluateResult(
>                      ^
> In file included from /Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:19:
> In file included from /Users/minhancao/velox/./velox/common/base/tests/GTestUtils.h:19:
> /Users/minhancao/velox/_build/debug/_deps/gtest-src/googletest/include/gtest/gtest.h:1374:11: error: invalid operands to binary expression ('const char[17]' and 'const std::shared_ptr<std::string>')
>   if (lhs == rhs) {
>       ~~~ ^  ~~~
> /Users/minhancao/velox/_build/debug/_deps/gtest-src/googletest/include/gtest/gtest.h:1393:12: note: in instantiation of function template specialization 'testing::internal::CmpHelperEQ<char[17], std::shared_ptr<std::string>>' requested here
>     return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
>            ^
> /Users/minhancao/velox/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2899:5: note: in instantiation of function template specialization 'testing::internal::EqHelper::Compare<char[17], std::shared_ptr<std::string>, nullptr>' requested here
>     EXPECT_EQ("19:51:34.241 UTC", currentTimeStr);
> ```

minhancao commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1096,6 +1098,30 @@ struct ParseDateTimeFunction {
+    Timestamp ts = Timestamp::now();
+    ts.toTimezone(*sessionTimeZone_);
+
+    std::string tsString = ts.toStringTimeWithTimezone();
```
@aditi-pandit I'm not sure what you mean by this question, can you further elaborate on what you mean?

minhancao commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1096,6 +1098,30 @@ struct ParseDateTimeFunction {
+    sessionTimeZone_ = getTimeZoneFromConfig(config);
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
```
Done

pedroerp commented on velox/type/Timestamp.h:
```diff
@@ -178,6 +178,29 @@ struct Timestamp {
     return oss.str();
   }
 
+  std::string toStringTimeWithTimezone() const {
```
could we refactor this code in a way that we don't need to duplicate the function above?

aditi-pandit replied:
Yeah, it would be sufficient to add an optional parameter to the above function to print timezone, and then invoke it here with nanoprecision and print timezone for the current behavior.

minhancao replied:
@aditi-pandit @pedroerp I have reimplemented toString() to also handle the time with timezone format, please let me know if I need to change anything, thanks!

minhancao commented:
@aditi-pandit @pedroerp I have cleaned up my code and added testing for my function. Please review at your earliest convenience, thanks!

minhancao commented:
I have taken a look at the failing test for the ci/circleci: linux-build below, not sure what caused this failure, if there is any insight please let me know:

```
I20230522 19:51:21.899892 304011 ExpressionFuzzer.cpp:1240] ==============================> Started iteration 1881 (seed: 1241120905)
I20230522 19:51:21.900439 304011 ExpressionVerifier.cpp:88] Executing expression: get_current_time()
I20230522 19:51:21.900451 304011 ExpressionVerifier.cpp:31] 0 vectors as input:
I20230522 19:51:21.900789 304011 ExpressionVerifier.cpp:68] All results match.
I20230522 19:51:21.900825 304011 ExpressionVerifier.cpp:88] Executing expression: get_current_time()
I20230522 19:51:21.900836 304011 ExpressionVerifier.cpp:31] 0 vectors as input:
E20230522 19:51:21.901139 304011 Exceptions.h:68] Line: ../../velox/expression/tests/ExpressionVerifier.cpp:61, Function:compareVectors, Expression: left->equalValueAt(right.get(), i, i) Different results at idx '1': '19:51:21.900 GMT' vs. '19:51:21.901 GMT', Source: RUNTIME, ErrorCode: INVALID_STATE
I20230522 19:51:21.901840 304011 ExpressionVerifier.cpp:316] Persisted input: --fuzzer_repro_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn --input_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn/input_vector --result_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn/result_vector --sql_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn/sql
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Different results at idx '1': '19:51:21.900 GMT' vs. '19:51:21.901 GMT'
Retriable: False
Expression: left->equalValueAt(right.get(), i, i)
Function: compareVectors
File: ../../velox/expression/tests/ExpressionVerifier.cpp
Line: 61
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZN8facebook5velox4test12_GLOBAL__N_114compareVectorsERKSt10shared_ptrINS0_10BaseVectorEES7_
# 6  _ZN8facebook5velox4test18ExpressionVerifier6verifyERKSt10shared_ptrIKNS0_4core10ITypedExprEERKS3_INS0_9RowVectorEEOS3_INS0_10BaseVectorEEbSt6vectorIjSaIjEE
# 7  _ZN8facebook5velox4test16ExpressionFuzzer2goEv
# 8  _ZN8facebook5velox4test16expressionFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIPKNS0_4exec17FunctionSignatureESaISD_EESt4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_SF_EEEm
# 9  _ZN12FuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES7_
# 10 main
# 11 __libc_start_main
# 12 _start
```

aditi-pandit commented on velox/type/Timestamp.h:
```diff
@@ -148,7 +148,8 @@ struct Timestamp {
   std::string toString(
-      const Precision& precision = Precision::kNanoseconds) const {
+      const Precision& precision = Precision::kNanoseconds,
+      int timeWithTimezoneFormatFlag = 0) const {
```
This can be bool variable to make it cleaner.

Maybe call the variable printTimezone.

minhancao replied:
Done

aditi-pandit commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -2858,6 +2866,17 @@ TEST_F(DateTimeFunctionsTest, dateParse) {
+  auto emptyRowVector = makeRowVector(ROW({}), 1);
+  auto tz = "America/Los_Angeles";
+  setQueryTimeZone(tz);
+  auto currentTimeBefore = getCurrentTime(tz);
```
Nit : Call this variable before is sufficent.

minhancao replied:
Done

aditi-pandit commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -2858,6 +2866,17 @@ TEST_F(DateTimeFunctionsTest, dateParse) {
+  auto tz = "America/Los_Angeles";
+  setQueryTimeZone(tz);
+  auto currentTimeBefore = getCurrentTime(tz);
+  auto result = evaluateOnce<std::string>("current_time", emptyRowVector);
```
Does emptyRowVector need to be a variable ? Can the empty row be constructed in place here ?

minhancao replied:
Changed it to be constructed in place in the function call

aditi-pandit commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -2858,6 +2866,17 @@ TEST_F(DateTimeFunctionsTest, dateParse) {
+  setQueryTimeZone(tz);
+  auto currentTimeBefore = getCurrentTime(tz);
+  auto result = evaluateOnce<std::string>("current_time", emptyRowVector);
+  auto currentTimeAfter = getCurrentTime(tz);
```
Rename this variable 'after'.

minhancao replied:
Done

aditi-pandit commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1096,6 +1098,29 @@ struct ParseDateTimeFunction {
+struct CurrentTimeFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  const date::time_zone* sessionTimeZone_ = nullptr;
```
This local variable should not end in "_". Please remove it.

minhancao replied:
Done

minhancao commented:
@aditi-pandit I have implemented your PR comments, please take a look

aditi-pandit commented:
> I have taken a look at the failing test for the ci/circleci: linux-build below, not sure what caused this failure or how I should fix it since the current time is always changing... if there is any insight please let me know:
> 
> ```
> I20230522 19:51:21.899892 304011 ExpressionFuzzer.cpp:1240] ==============================> Started iteration 1881 (seed: 1241120905)
> I20230522 19:51:21.900439 304011 ExpressionVerifier.cpp:88] Executing expression: get_current_time()
> I20230522 19:51:21.900451 304011 ExpressionVerifier.cpp:31] 0 vectors as input:
> I20230522 19:51:21.900789 304011 ExpressionVerifier.cpp:68] All results match.
> I20230522 19:51:21.900825 304011 ExpressionVerifier.cpp:88] Executing expression: get_current_time()
> I20230522 19:51:21.900836 304011 ExpressionVerifier.cpp:31] 0 vectors as input:
> E20230522 19:51:21.901139 304011 Exceptions.h:68] Line: ../../velox/expression/tests/ExpressionVerifier.cpp:61, Function:compareVectors, Expression: left->equalValueAt(right.get(), i, i) Different results at idx '1': '19:51:21.900 GMT' vs. '19:51:21.901 GMT', Source: RUNTIME, ErrorCode: INVALID_STATE
> I20230522 19:51:21.901840 304011 ExpressionVerifier.cpp:316] Persisted input: --fuzzer_repro_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn --input_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn/input_vector --result_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn/result_vector --sql_path /tmp/fuzzer_repro/velox_expressionVerifier_RPtwXn/sql
> terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
>   what():  Exception: VeloxRuntimeError
> Error Source: RUNTIME
> Error Code: INVALID_STATE
> Reason: Different results at idx '1': '19:51:21.900 GMT' vs. '19:51:21.901 GMT'
> Retriable: False
> Expression: left->equalValueAt(right.get(), i, i)
> Function: compareVectors
> File: ../../velox/expression/tests/ExpressionVerifier.cpp
> Line: 61
> Stack trace:
> # 0  _ZN8facebook5velox7process10StackTraceC1Ei
> # 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
> # 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
> # 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
> # 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
> # 5  _ZN8facebook5velox4test12_GLOBAL__N_114compareVectorsERKSt10shared_ptrINS0_10BaseVectorEES7_
> # 6  _ZN8facebook5velox4test18ExpressionVerifier6verifyERKSt10shared_ptrIKNS0_4core10ITypedExprEERKS3_INS0_9RowVectorEEOS3_INS0_10BaseVectorEEbSt6vectorIjSaIjEE
> # 7  _ZN8facebook5velox4test16ExpressionFuzzer2goEv
> # 8  _ZN8facebook5velox4test16expressionFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIPKNS0_4exec17FunctionSignatureESaISD_EESt4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_SF_EEEm
> # 9  _ZN12FuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES7_
> # 10 main
> # 11 __libc_start_main
> # 12 _start
> ```

Hi @minhancao, 
We could disable current_time in the Expression Fuzzer test. There is a list of skip_functions in ExpressionFuzzerTest.cpp
The rest of your code is looking good otherwise.

aditi-pandit commented:
> @aditi-pandit I have taken a look at presto documentation https://prestodb.io/docs/current/functions/datetime.html?highlight=current_time and I was wondering if I should reimplement and create new type named TimeWithTimeZoneType.h that should be similar to TimestampWithTimeZoneType.h? I can then implement the toString() in the new file TimeWithTimeZoneType.h to be the format that I want `HH:MM:SS.ZZZ UTC`. Should I go this route or stick with what I currently have implemented?
> 
> ```
> current_time -> time with time zone
>     Returns the current time as of the start of the query.
> ```

@minhancao : This is a good point you raised. The Presto function returns a time with timezone type. We need to add a new time with timezone type in Velox. There is an issue for this https://github.com/prestodb/presto/issues/18845. This implementation would be a bit involved.

Lets stay off this PR for now.
